File: python/sample_search/informed_rrt.py
'''
@file: informed_rrt.py
@breif: Informed RRT* motion planning
@author: Winter
@update: 2023.1.18
'''
import os, sys
import numpy as np
from functools import partial
import matplotlib.pyplot as plt

sys.path.append(os.path.abspath(os.path.join(__file__, "../../")))

from .rrt_star import RRTStar
from utils import Env, Node

class ellipse:
    '''
    Ellipse sampling.
    '''    
    @staticmethod
    def transform(a: float, c: float, p1: tuple, p2: tuple) -> np.ndarray:
        # center
        center_x = (p1[0] + p2[0]) / 2
        center_y = (p1[1] + p2[1]) / 2

        # rotation
        theta = - np.arctan2(p2[1] - p1[1], p2[0] - p1[0])

        # transform
        b = np.sqrt(a ** 2 - c ** 2)
        T = np.array([[ a * np.cos(theta), b * np.sin(theta), center_x],
                      [-a * np.sin(theta), b * np.cos(theta), center_y],
                      [                 0,                 0,        1]])
        return T

class InformedRRT(RRTStar):
    '''
    Class for Informed RRT* motion planning.
    [1] Optimal Sampling-based Path Planning Focused via Direct
        Sampling of an Admissible Ellipsoidal heuristic

    Parameters
    ----------
    start: tuple
        start point coordinate
    goal: tuple
        goal point coordinate
    env: Env
        environment
    max_dist: float
        Maximum expansion distance one step
    sample_num: int
        Maximum number of sample points
    r: float
        optimization radius
    goal_sample_rate: float
        heuristic sample

    Examples
    ----------
    >>> from utils import Map
    >>> from sample_search import InformedRRT
    >>> start = (5, 5)
    >>> goal = (45, 25)
    >>> env = Map(51, 31)
    >>> planner = InformedRRT(start, goal, env)
    >>> planner.run()
    '''
    def __init__(self, start: tuple, goal: tuple, env: Env, max_dist: float,
                 sample_num: int, r: float, goal_sample_rate: float = 0.05) -> None:
        super().__init__(start, goal, env, max_dist, sample_num, goal_sample_rate)
        # optimization radius
        self.r = r
        # best planning cost
        self.c_best = float("inf")
        # distance between start and goal
        self.c_min = self.dist(self.start, self.goal)
        # ellipse sampling
        self.transform = partial(ellipse.transform, c=self.c_min / 2, p1=start, p2=goal)
    
    def __str__(self) -> str:
        return "Informed RRT*"

    def plan(self):
        '''
        Informed-RRT* motion plan function.

        Return
        ----------
        cost: float
            path cost
        path: list
            planning path
        '''
        # generate a random node in the map
        node_rand = self.generateRandomNode()

        # visited
        if node_rand in self.sample_list:
            return 0, None

        # generate new node
        node_new = self.getNearest(self.sample_list, node_rand)
        if node_new:
            self.sample_list.append(node_new)
            dist = self.dist(node_new, self.goal)
            # goal found
            if dist <= self.max_dist and not self.isCollision(node_new, self.goal):
                self.goal.parent = node_new.current
                self.goal.g = node_new.g + self.dist(self.goal, node_new)
                self.sample_list.append(self.goal)
                return self.extractPath(self.sample_list)
        return 0, None

    def run(self) -> None:
        '''
        Running both plannig and animation.
        '''
        best_cost, best_path = float("inf"), None

        # main loop
        for i in range(self.sample_num):
            cost, path = self.plan()
            # update
            if path and cost < best_cost:
                self.c_best = best_cost = cost
                best_path = path
            # animation
            if  i % 30 == 0:
                self.animation(best_path, best_cost)
        plt.show()

    def generateRandomNode(self) -> Node:
        '''
        Generate a random node to extend exploring tree.

        Return
        ----------
        node: Node
            a random node based on sampling
        '''
        # ellipse sample
        if self.c_best < float("inf"):
            while True:
                # unit ball sample
                p = np.array([.0, .0, 1.])
                while True:
                    x, y = np.random.uniform(-1, 1), np.random.uniform(-1, 1)
                    if x ** 2 + y ** 2 < 1:
                        p[0], p[1] = x, y
                        break
                # transform to ellipse
                p_star = self.transform(self.c_best / 2) @ p.T
                if self.delta <= p_star[0] <= self.env.x_range - self.delta and \
                   self.delta <= p_star[1] <= self.env.y_range - self.delta:
                    return Node((p_star[0], p_star[1]), None, 0, 0)
        # random sample
        else:
            return super().generateRandomNode()

    def animation(self, path, cost):
        self.plot.clean()
        name = str(self) + "\ncost: " + str(cost)
        self.plot.plotEnv(name)
        for x in self.sample_list:
            if x.parent:
                plt.plot([x.parent[0], x.current[0]], [x.parent[1], x.current[1]], 
                    color="#dddddd", linestyle="-")
        if self.c_best < float("inf"):
            self.drawEllipse()
        if path:
            self.plot.plotPath(path)
        plt.pause(0.01)

    def drawEllipse(self):
        t = np.arange(0, 2 * np.pi + 0.1, 0.1)
        x = [np.cos(it) for it in t]
        y = [np.sin(it) for it in t]
        z = [1 for _ in t]
        fx = self.transform(self.c_best / 2) @ np.array([x, y, z])
        plt.plot(fx[0, :], fx[1, :], linestyle='--', color='darkorange', linewidth=2)

File: matlab/sample_search/informed_rrt.m
function [path, flag, cost, expand] = informed_rrt(map, start, goal)
%%
% @file: informed_rrt.m
% @breif: Informed RRT* motion planning
% @paper: Optimal Sampling-based Path Planning Focused via Direct
%                   Sampling of an Admissible Ellipsoidal heuristic
% @author: Winter
% @update: 2023.2.2

%%
     % optimal radius
    param.r = 10;
    % Maximum expansion distance one step
    param.max_dist = 0.5;
    % Maximum number of sample points
    param.sample_num = 5500;
    % heuristic sample
    param.goal_sample_rate = 0.05;
    % map size
    [param.x_range, param.y_range] = size(map);
    % resolution
    param.resolution = 0.1;
    % big number
    param.big_num = 10000;
    % best planning cost
    param.c_best = param.big_num;
    % distance between start and goal
    param.c_min = dist(start, goal');
    
    % sample list
    sample_list  = [start, 0, start];
    
    path = [];
    flag = false;
    cost = 0;
    expand = [];
    
    % main loop
    for i=1: param.sample_num
        [cost_, flag, sample_list, path_] = plan(sample_list, start, goal, map, param);
        if flag && cost < param.c_best
            param.c_best = cost_;
            path = path_;
            cost = cost_;
        end
    end
    
    expand = sample_list;
end

%%
function index = loc_list(node, list, range)
% @breif: locate the node in given list
    num = size(list);
    index = 0;
    if ~num(1)
        return
    else  
        for i=1:num(1)
            if isequal(node(range), list(i, range))
                index = i;
                return;
            end
        end
    end
end

function [cost, flag, node_list, path] = plan(node_list, start, goal, map, param)
    cost = 0;
    flag = false;
    path = [];

    % generate a random node in the map
    node_rand = generate_node(start, goal, param);

    % visited
    if loc_list(node_rand, node_list, [1, 2])
        return
    end

    % generate new node
    [node_new, success] = get_nearest(node_list, node_rand, map, param);
    if success
        node_list = [node_new; node_list];
        distance = dist(node_new(1:2), goal');

        % goal found
        if distance <= param.max_dist && ~is_collision(node_new(1:2), goal, map, param)
            goal_ = [goal, node_new(3) + distance, node_new(1:2)];
            node_list = [goal_; node_list];
            flag = true;
            cost = goal_(3);
            path = extract_path(node_list, start);
            node_list(1, :) = [];
            return
        end
    end
end

function node = generate_node(start, goal, param)
%breif: Generate a random node to extend exploring tree.
    % ellipse sample
    if param.c_best < param.big_num
        while true
            % unit ball sample
            p = [0, 0, 1];
            while true
                x = -1 + 2 * rand();
                y = -1 + 2 * rand();
                if x * x + y * y < 1
                    p(1) = x; p(2) = y;
                    break
                end
            end
            % transform to ellipse
            p_star = transform(param.c_best / 2, param.c_min / 2, start, goal) * p';
            if 0 <= p_star(1) <= param.x_range && 0 <= p_star(2) <= param.y_range
                node = [p_star(1), p_star(2)];
                return;
            end
        end
    % random sample
    else
        if rand() > param.goal_sample_rate
            x = 0.5 + (param.x_range - 1) * rand();
            y = 0.5 + (param.y_range - 1) * rand();
            node = [x, y];
            return
        end

        node = goal;
        return
   end
end

function [new_node, flag] = get_nearest(node_list, node, map, param)
%@breif: Get the node from `node_list` that is nearest to `node`.
    flag = false;
    % find nearest neighbor
    dist_vector = dist(node_list(:, 1:2), node');
    [~, index] = min(dist_vector);
    node_near = node_list(index, :);

    % regular and generate new node
    distance = min(dist(node_near(1:2), node'), param.max_dist);
    theta = angle(node_near, node);
    new_node = [node_near(1) + distance * cos(theta), ...
                           node_near(2) + distance * sin(theta), ...
                           node_near(3) + distance, ...
                           node_near(1:2)];

    % obstacle check
    if is_collision(new_node(1:2), node_near(1:2), map, param)
        return
    end
    
    %  rewire optimization
    [node_num, ~] = size(node_list);
    for i=1:node_num
        node_n = node_list(i, :);
        %  inside the optimization circle
        new_dist = dist(node_n(1:2), new_node(1:2)');
        if new_dist < param.r
            cost = node_n(3) + new_dist;
            %  update new sample node's cost and parent
            if new_node(3) > cost && ~is_collision(new_node(1:2), node_n(1:2), map, param)
                new_node(4:5) = node_n(1:2);
                new_node(3) = cost;
            else
                %  update nodes' cost inside the radius
                cost = new_node(3) + new_dist;
                if node_n(3) > cost && ~is_collision(new_node(1:2), node_n(1:2), map, param)
                    node_list(i, 4:5) = new_node(1:2);
                    node_list(i, 3) = cost;
                end
            end
        else
            continue;
        end
    end
    
    flag = true;
end

function flag = is_collision(node1, node2, map, param)
%@breif: Judge collision when moving from node1 to node2.
    flag = true;
    theta = angle(node1, node2);
    distance = dist(node1, node2');

      % distance longer than the threshold
      if (distance > param.max_dist)
          return
      end
        
      % sample the line between two nodes and check obstacle
      n_step = round(distance / param.resolution);
      for i=1:n_step
          x = node1(1) + i * param.resolution * cos(theta);
          y = node1(2) + i * param.resolution * sin(theta);
          if map(round(x), round(y)) == 2
              return
          end
      end
      
      flag = false;
end

function path = extract_path(close, start)
% @breif: Extract the path based on the CLOSED set.
    path  = [];               
    closeNum = length(close(:, 1));
    index = 1;
    while 1
        path = [path; close(index, 1:2)];
        if isequal(close(index, 1:2), start)   
            break;
        end
        for i=1:closeNum
            if isequal(close(i, 1:2), close(index, 4:5))
                index = i;
                break;
            end
        end
    end
end

 function theta = angle(node1, node2)
    theta = atan2(node2(2) - node1(2), node2(1) - node1(1));
 end

function T = transform(a, c, start, goal)
    % center
    center_x = (start(1) + goal(1)) / 2;
    center_y = (start(2) + goal(2)) / 2;

    % rotation
    theta =  -angle(start, goal);

    % transform
    b = sqrt(a * a - c * c);
    T = [ a * cos(theta), b * sin(theta),  center_x; ...
            -a * sin(theta), b * cos(theta), center_y; ...
                                 0,                      0,            1];
end